include/ibamr/ComplexFluidForcing.h

```diff
@@ -0,0 +1,330 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef included_ComplexFluidForcing
+#define included_ComplexFluidForcing
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "IBAMR_config.h"
+
+#include "ibamr/AdvDiffComplexFluidConvectiveOperator.h"
+#include "ibamr/AdvDiffSemiImplicitHierarchyIntegrator.h"
+#include "ibamr/CFGiesekusRelaxation.h"
+#include "ibamr/CFOldroydBRelaxation.h"
+#include "ibamr/CFRoliePolyRelaxation.h"
+#include "ibamr/INSStaggeredHierarchyIntegrator.h"
+#include "ibamr/ibamr_enums.h"
+#include "ibamr/namespaces.h" // IWYU pragma: keep
+
+#include "ibtk/CartGridFunction.h"
+#include "ibtk/HierarchyGhostCellInterpolation.h"
+#include "ibtk/ibtk_utilities.h"
+#include "ibtk/muParserCartGridFunction.h"
+#include "ibtk/muParserRobinBcCoefs.h"
+
+#include "Box.h"
+#include "CartesianPatchGeometry.h"
+#include "CellData.h"
+#include "CellIndex.h"
+#include "CellVariable.h"
+#include "HierarchyDataOpsManager.h"
+#include "Index.h"
+#include "IntVector.h"
+#include "Patch.h"
+#include "PatchGeometry.h"
+#include "PatchHierarchy.h"
+#include "PatchLevel.h"
+#include "SideData.h"
+#include "Variable.h"
+#include "VariableContext.h"
+#include "VariableDatabase.h"
+#include "VisItDataWriter.h"
+#include "tbox/Database.h"
+#include "tbox/Pointer.h"
+
+#include <Eigen/Cholesky>
+#include <Eigen/Core>
+#include <Eigen/Eigenvalues>
+#include <unsupported/Eigen/MatrixFunctions>
+
+#include <math.h>
+#include <mpi.h>
+
+#include <string>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class ComplexFluidForcing provides an interface for specifying a viscoelastic stress to be added to the
+ * Navier-Stokes equations. The class uses the advection diffusion integrator to update the viscoelastic stress.
+ *
+ * One can choose from the pre-programmed models Oldroyd-B, Giesekus, or Rolie-Poly. Optionally, you can also register a
+ * relaxation operator that computes the relaxation function. This class currently solves for the conformation tensor or
+ * the square root or logarithm of the conformation tensor. THe current assumption is that the stress is linearly
+ * related to the conformation tensor through the elastic modulus.
```

Comment:
Is it possible to relax this requirement in a subclass, or would it require more substantial modifications?

Reply:
The goal was to allow a function to be registered that does this transformation automatically. However, nobody was particularly interested in models that aren't linearly related, so I never got around to it. It should be a relatively painless change. There are a couple of `TODO:` comments in the code where this assumption is made.